sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+
+_Returns_: The default memory pool associated with the context and `dev` for
+allocating with the allocation kind `usm::device`.
+
+
+=== Asynchronous malloc & free
+
+This extension introduces a series of new enqueue functions for enqueueing
+asynchornous malloc and free commands which operate with the memory pools also
+introduced in this extension.
+
+All enqueue functions introduced have overloads which take a SYCL `queue` and a
+SYCL `handler`. None of enqueue functions return a SYCL `event` direction, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+void *async_malloc(queue q, size_t size);
+
+void *async_malloc(handler h, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+----
+
+_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`
+which will asynchornously allocate memory of size `size` in bytes, allocating
+from the memory pool `pool` if provided, otherwise allocation from the default
+memory pool associated with the SYCL context and device associated with `q` or
+`h`. Memory is first allocated from the memory pool if possible, otherwise
+memory is allocated from the OS to the memory pool to provide enough memory in
+the memory pool for the allocation.
+
+_Returns_: A pointer to the address of a memory reservation.
+
+_Throws_: An exception with the `errc::invalid` error code if `size` is zero.
+An exception with the `errc::memory_allocation` error code if the allocation
+brings the memory pool over the maximum size. This error can be thrown
+asynchornously.
+
+[_Note:_ Accessing the memory at the address of the pointer returned by
+asynchornous malloc functions before the command has completed execution is
+undefined behaviour. _{endnote}_]
+
+[source, c++]
+----
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+----
+
+_Effects_: Enqueues a commands to `q` or the SYCL queue associated with `h`
+which will asynchornously free the memory allocation at the address of `ptr`.
+Memory will be freed from the memory pool to be used by other asynchronous
+malloc commands which execute later, and will not free to the OS until the SYCL
+queue associated with the asycnhronous allocation command has been synchornized
+with.
```

Comment:
```suggestion
_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h`
which will asynchronously free the memory allocation at the address of `ptr`.
Memory will be freed from the memory pool to be used by other asynchronous
malloc commands which execute later, and will not free to the OS until the SYCL
queue associated with the asynchronous allocation command has been synchronized
with.
```

Reply:
Fixed in https://github.com/intel/llvm/pull/14800/commits/2c51adf28ece21c0e91c2c0ec85440ae1153e536.